### DISCUSSION TITLE
The MMDloss generates negative outputs

### DISCUSSION THREAD

**Discussion by lyfeng001** [KevinMusgrave/pytorch-adapt#103] [Q&A] [answered]
I initialized the MMDLoss class and used it to process my data. Here is the code:

```
mmd_loss = `MMDLoss(mmd_type="quadratic")
……
loss_mmd = mmd_loss(source_embeddings.view(-1, 167*768), target_embeddings.view(-1, 167*768))
```

I expected the MMD loss to output a positive value, but some outputs are negative during training. Could you help me understand why this might be happening?

![3449eea9-837b-4601-8587-b1f861cee7e1](https://github.com/user-attachments/assets/1300861a-3e40-4624-86d8-19ec86d18a35)

the blue one is the training loss, it became negative during the training.

**Comment by KevinMusgrave** [comment, marked as answer]
I can't remember why, but here's an explanation from the official implementation of MMD:

"The value of the mmd loss could be negative since we used the linear-time unbiased estimate of the mmd, which lends us an O(n) time cost but may cause negative mmd values for some mini-batch. This negative mmd value however will not influence the final classification performance."

Link: https://github.com/thuml/DAN/blob/master/README.md

**Reply by lyfeng001** [reply]
I get it! Thanks for your reply!
